src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamValueTaskSource.cs

```diff
@@ -0,0 +1,252 @@
+                {
+                    packedResult = ((ulong)FileStreamHelpers.ResultError | errorCode);
+                }
+                else
+                {
+                    packedResult = ((ulong)FileStreamHelpers.ResultSuccess | numBytes);
+                }
+
+                // Stow the result so that other threads can observe it
+                // And, if no other thread is registering cancellation, continue
+                if (Interlocked.Exchange(ref valueTaskSource._result, (long)packedResult) == FileStreamHelpers.NoResult)
+                {
+                    // Successfully set the state, attempt to take back the callback
+                    if (Interlocked.Exchange(ref valueTaskSource._result, FileStreamHelpers.CompletedCallback) != FileStreamHelpers.NoResult)
+                    {
+                        // Successfully got the callback, finish the callback
+                        valueTaskSource.CompleteCallback(packedResult);
+                    }
+                    // else: Some other thread stole the result, so now it is responsible to finish the callback
+                }
+                // else: Some other thread is registering a cancellation, so it *must* finish the callback
+            }
+
+            private void CompleteCallback(ulong packedResult)
+            {
+                // Free up the native resource and cancellation registration
+                CancellationToken cancellationToken = _cancellationRegistration.Token; // access before disposing registration
+                ReleaseNativeResource();
+
+                // Unpack the result and send it to the user
+                long result = (long)(packedResult & FileStreamHelpers.ResultMask);
+                if (result == FileStreamHelpers.ResultError)
+                {
+                    int errorCode = unchecked((int)(packedResult & uint.MaxValue));
+                    if (errorCode == Interop.Errors.ERROR_OPERATION_ABORTED)
+                    {
+                        _source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException(cancellationToken.IsCancellationRequested ? cancellationToken : new CancellationToken(canceled: true))));
+                    }
+                    else
+                    {
+                        Exception e = Win32Marshal.GetExceptionForWin32Error(errorCode);
+                        e.SetCurrentStackTrace();
+                        _source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(e));
+                    }
+                }
+                else
+                {
+                    Debug.Assert(result == FileStreamHelpers.ResultSuccess, "Unknown result");
+                    _source.SetResult((int)(packedResult & uint.MaxValue) + _numBufferedBytes);
+                }
+            }
+
+            private void Cancel(CancellationToken token)
+            {
```

Comment:
Missing comment and asserts:

https://github.com/dotnet/runtime/blob/57bbdee53593fb89913c5f2e19fdcdc10eac3980/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamCompletionSource.Win32.cs#L211-L215

Reply:
The first assert is not needed because the method argument is not an `object? state`.

I'll add the assert that checks for overlapped an completed status.

Reply:
The `state` param could be added to this method and allows you to conserve the first Debug.Assert.